tifs-a-tikv-based-partition-tolerant-strictly-consistent-file-system.md

```diff
@@ -0,0 +1,326 @@
+* There's no complicated SDK or API for space management. You simply call the file system API or run a shell script.
+* TiFS lets you take a standalone application that supports multi-instance collaboration and turn it into a distributed application. For example, **SQLite + TiFS = yet another distributed relational database**. Admittedly, to use TiFS this way, the application cannot rely on page cache or other caching mechanisms to avoid writes being invisible.
+
+![TiFS makes SQLite another distributed relational database](media/tifs-makes-sqlite-another-distributed-relational-database.gif)
+<div class="caption-center">TiFS makes SQLite another distributed relational database</div>
+
+## Tests and benchmark
+
+During the Hackathon, we used [pjdfstest](https://github.com/pjd/pjdfstest) to test TiFS's correctness. But because pjdfstest doesn't cover read/write correctness or concurrency correctness, we will need to add other tests.
+
+Theoretically, there are three major factors that influence the read/write performance of TiFS: the size of file system blocks, the network latency, and the size of load blocks. Here we'll show the [benchmarking results](https://github.com/Hexilee/tifs.benchmark) of the read/write IOPS and speed.
+
+### IOPS
+
+> Note: TiKV is a sophisticated system, in which there are logical duration, disk I/O duration, and network duration. In this article, we simplify TiKV to a single replica for demonstration purposes.
+
+Let's first take a look at IOPS. Because sequential reads and writes perform I/O operations linearly, each I/O operation is a transaction in TiKV. If we ignore the minor differences between each operation, the duration of an I/O operation, **_T_**, is the reciprocal of **_IOPS_**. In addition, if we don't count stream processing, we can consider **_T_** as the linear addition of the following four variables:
+
+* **_T<sub>f</sub>_**: the I/O duration of [FUSE](https://en.wikipedia.org/wiki/Filesystem_in_Userspace).
+* **_T<sub>c</sub>_**: the logical duration of TiFS.
+* **_T<sub>n</sub>_** : the network transmission time.
+* **_T<sub>s</sub>_** : the logical duration of TiKV.
+
+Accordingly, we have this equation:
+
+![TiFS IOPS Equation](media/tifs-iops-equation.png)
+
+#### Read IOPS
+
+For read operations, **_T<sub>f</sub>_** is positively correlated to the size of load blocks. The size of data read or written by TiFS per I/O operation must be an integer multiple of the size of file system blocks. Therefore, **_T<sub>n</sub>_** and **_T<sub>s</sub>_** are positively correlated to the larger value between load blocks and file system blocks; however, with larger traffic, network and disk I/O may take longer. **_T<sub>c</sub>_** is unknown.
+
+**The following diagram shows how the sequential read IOPS varies with the size of load blocks.** The four lines represent different filesystem block sizes and data replicas.
+
+![Read IOPS benchmarking results](media/tifs-read-iops-benchmarking-results.png)
+
```

Comment:
Please add an image caption.

Reply:
Image captions are good for SEO, because they help the search engine understand a website's content.